### PR TITLE
Namespace Program

### PR DESCRIPTION
**Pull request by jackrieck**
Namespace Gatekeepers
- Namespace Gatekeepers are initialized by the program with `create_namespace_gatekeeper`, they start with 0 filters
- Filters are appended and removed via their respective instructions
- Currently the filter impl for Category and Key only use the namespace and mint fields respectively, what is the vision for these 2 filters?
- Does an artifact need to pass all filters to be included in the namespace or just 1?
- PermissivenessSettings are set at `initialize_namespace` but really are used with the Gatekeeper, I wonder if it makes more sense to set them at the gatekeeper level?

Artifact Caching
- An artifact must be added to a namespace via `join_namespace` to be eligible to be cached, an item cannot leave a namespace via `leave_namespace` without first being removed from the namespace cache
- There are 2 new fields in the `Namespace` Account, `max_pages` and `full_pages`, `max_pages` denotes how many pages a namespace is allowed to use for caching, this is currently hardcoded to 10. `full_pages` keeps track of which pages in the cache are full and cannot accept any more artifacts.
- The program will use `full_pages` to determine which page or index needs to be passed in via the client, this is so we always fill up the lower pages first and use all the space at our disposal. The client can query the artifact's account to figure out which page it's cached, this is useful to remove it from the cache as you will need to pass in the appropriate account.
- For an artifact to be joined and cached to a particular namespace, that artifact needs to expose 4 instructions for `join_namespace`, `leave_namespace`, `cache_artifact` and `uncache_artifact` For now I have just done `item_class` in the `item` program to test this out.

Client Side
- CLI stubbed out but not tested, SDK used by CLI and tests

Testing
- I need the keypairs for the programs to test in dev/main but I have setup local tests that run agains the validator simulation, these are runnable with `anchor test --skip-lint`
- Right now really only the happy paths are tested

Notes/Questions
- As we begin to do CPI and share types between programs, it may be a good idea to have another crate that is simply all the shareable types so that we dont get into dependency cycle issues. In addition we have a lot of similar helper functions in utils, these could be added there as well.
- Namespace Program as its written assumes no program which wants to be namespaced will bring in the Namespace program as a dependency as this creates a cycle.
- Seems like the raindrops typescript library uses an older version of @solana/spl-token, we should upgrade this

Next Steps
- If this PR and methods pertaining to caching and joining are acceptable, the next things would be to implement the instructions for the other crates and the rest of the item artifacts.
- More extensive testing to cover all the edge cases
- Add the necessary changes so that all programs pass anchor linting
- Add Vaults to accept $RAIN and other tokens as payment for joining namespaces
- CLI and test

### CODE REVIEW

File: js/lib/src/state/namespace.ts
Comment (rufflian) [inline js/lib/src/state/namespace.ts:31]: Do we want to throw if an invalid permissiveness is passed in?

File: js/lib/src/state/namespace.ts
Comment (rufflian) [inline js/lib/src/state/namespace.ts:48]: What's the purpose of using `var` instead of `const` here?

**Review by rufflian — COMMENTED** [review]
Looks good! I've only gone through the lib code so far (I'm afraid rust isn't my speciality)

File: rust/item/src/lib.rs
Comment (cotemper) [inline rust/item/src/lib.rs:31]: Can we keep the old program IDs?

File: rust/item/src/lib.rs
Comment (cotemper) [inline rust/item/src/lib.rs:1746]: Educational question, what is `ns.index` when leaving namespace?

File: rust/item/src/utils.rs
Comment (cotemper) [inline rust/item/src/utils.rs:1817]: What is the difference between `get_instruction_relative` and this?

File: rust/namespace/src/lib.rs
Comment (cotemper) [inline rust/namespace/src/lib.rs:23]: Can we update the MATCH_ID with the correct program ID?

File: rust/namespace/src/utils.rs
Comment (cotemper) [inline rust/namespace/src/utils.rs:280]: What can `Artifact` be? Is `ItemClass` also `Artifact`?

**Review by cotemper — CHANGES_REQUESTED** [review]

File: js/lib/src/constants/programIds.ts
Comment (jackrieck) [reply js/lib/src/constants/programIds.ts:8]: I dont have the keypairs for the programs so I wasnt able to use those while testing, when i get them ill change this back before merging

File: js/lib/src/constants/namespace.ts
Comment (jackrieck) [reply js/lib/src/constants/namespace.ts:2]: The reason I was changing this is because we end up using `as` when importing multiple `PREFIX` from different libraries

File: js/lib/src/constants/programIds.ts
Comment (jackrieck) [reply js/lib/src/constants/programIds.ts:9]: I will not end up changing these, I just dont have the keypairs so I was testing with keypairs I had locally.

File: js/lib/src/instructions/namespace.ts
Comment (jackrieck) [reply js/lib/src/instructions/namespace.ts:98]: I think both ways work, was just trying to make it convenient to the caller

File: js/lib/src/instructions/namespace.ts
Comment (jackrieck) [reply js/lib/src/instructions/namespace.ts:211]: I think both ways work, was just trying to make it convenient to the caller

File: js/lib/src/instructions/namespace.ts
Comment (jackrieck) [reply js/lib/src/instructions/namespace.ts:413]: Good question, I'll look into this.

File: rust/item/src/lib.rs
Comment (jackrieck) [reply rust/item/src/lib.rs:31]: Once I get the keypairs I will revert this change, its only temporary and I wont merge the PR until I do

File: rust/item/src/lib.rs
Comment (jackrieck) [reply rust/item/src/lib.rs:1746]: When an artifact joins a namespace it is not cached by default, so index will remain `None`. When `cache_artifact` is called, this will change to the index number in which it is cached. There are 10 indexes per namespace each can hold 100 pubkeys, we do not allow an artifact which is still cached to leave a namespace.

File: rust/item/src/utils.rs
Comment (jackrieck) [reply rust/item/src/utils.rs:1817]: I'm not sure, this is how I've always done it, but ill check out that function.

File: rust/namespace/src/utils.rs
Comment (jackrieck) [reply rust/namespace/src/utils.rs:280]: Yeah, anything that is namespace-able is an artifact. So far I have only implemented ItemClass, if everyone is fine with how I did it i'll add the others

File: rust/item/src/lib.rs
Comment (jackrieck) [reply rust/item/src/lib.rs:31]: I dont think the anchor provider is what I need to use right? https://github.com/raindrops-protocol/raindrops/blob/4f1fd9c8847ae2f6da5a45a2ea6a743d084b2195/js/lib/src/state/common.ts#L50

File: rust/item/src/utils.rs
Comment (jackrieck) [reply rust/item/src/utils.rs:1817]: From the docs it looks like they're nearly identical, one is relative to current ix, one is absolute index in regards to entire tx

File: js/lib/src/state/namespace.ts
Comment (jackrieck) [reply js/lib/src/state/namespace.ts:31]: I dont think `AnchorPermissivenessType` is exactly correct, I need different enum for this one. https://github.com/raindrops-protocol/raindrops/blob/4f1fd9c8847ae2f6da5a45a2ea6a743d084b2195/js/lib/src/state/common.ts#L50

File: rust/item/src/utils.rs
Comment (cotemper) [reply rust/item/src/utils.rs:1817]: nvm, don't bother merge.

File: js/lib/src/constants/namespace.ts
Comment (cotemper) [reply js/lib/src/constants/namespace.ts:2]: If so, it's better to update other prefixes as well.

**Review by cotemper — APPROVED** [review]
LGTM!

**Pull request merged**
